Code generation

maksym-nazarenko opened the pull request:
Closes #47 

This PR replaces #104.

Complete before merge:
- ~[ ] errors typing~ (don't have a clear vision, so skipping atm)
- [x] use defined types to represent struct field types (currently, they are literal strings)

### What is the value of this PR?

It introduces new tool `mikrotik-codegen`, that supports `codegen` struct tag to generate Terraform resource based of struct definition in `client/` file.

The main purpose of this tool is to provide a fast and easy way to start working on new resource.

> **Warning**
>
> This tool is not supposed to be used as automated way to generate resource without human review.
> It just creates >90% of boilerplate code.


### How to use new feature?
Example usage might look like:
```go
type MikrotikResource struct{
	Id             string   `mikrotik:".id"             codegen:"id,mikrotikID,deleteID"`
	Name           string   `mikrotik:"name"            codegen:"name,required,terraformID"`
	Enabled        bool     `mikrotik:"enabled"         codegen:"enabled"`
	Items          []string `mikrotik:"items"           codegen:"items,elemType=string"`
	UpdatedAt      string   `mikrotik:"updated_at"      codegen:"updated_at,computed"`
	Unused         int      `mikrotik:"unused"          codegen:"-"`
	NotImplemented int      `mikrotik:"not_implemented" codegen:"not_implemented,omit"`
	Comment        string   `mikrotik:"comment"         codegen:"comment"`
}
```

with further code generation:
```sh
$ go run ./cmd/mikrotik-codegen -src client/resource.go -struct MikrotikResource > mikrotik/resource_new.go
```

### Next steps
As further improvement for this codebase, I see these useful features:
* Generate `client` API resource as well (based on struct definition only)
* Allow generating different parts of code: ` .... -only={client,client_test,terraform_resource,terraform_test}`

ddelnano commented on mikrotik/resource_bridge_test.go:
```diff
@@ -51,7 +51,7 @@ func testAccBridgeExists(resource string, record *client.Bridge) resource.TestCh
 		}
 		c := client.NewClient(client.GetConfigFromEnv())
-		remoteRecord, err := c.FindBridge(r.Primary.ID)
+		remoteRecord, err := c.FindBridge(r.Primary.Attributes["name"])
```
I assume this is unrelated to the rest of the changes?

maksym-nazarenko replied:
yes, I was testing the codegen on bridge resource and found that `ID` of the resource in state will be `.id` once we move to `PluginFramework`, so we should be more explicit in tests in the future.
I'll revert it for now.

ddelnano reviewed (commented):
Thanks for revamping this and I'm very excited to have this capability soon! I haven't reviewed everything yet (inner part of the code generation to go still).

I should be able to get through the rest of it this week. I intend to go through workflow end to end as I review the rest of this as well. I think I need to see it in action in order to effectively review the more involved parts of this change.

maksym-nazarenko commented on cmd/mikrotik-codegen/main.go:
```diff
@@ -0,0 +1,129 @@
+		destFile   = flag.String("dest", "-", "File to write result to. Default: write to stdout.")
+		srcFile    = flag.String("src", "", "Source file to parse struct from.")
+		structName = flag.String("struct", "", "Name of a struct to process.")
+		formatCode = flag.Bool("formatCode", true, "Whether to format resulting code.")
```
I am using it for debugging.
`Go` can't format code with syntax errors, and `expecting ',', got ')' at mikrotik/resource_new.go:103:15` says nothing, because line numbers do not correspond to template line numbers 🤷

ddelnano replied:
ah, I see. Can you add that additional detail in the flag description? I didn't understand how useful the option could be (I see that debugging use case as very important).

maksym-nazarenko commented on cmd/mikrotik-codegen/main.go:
```diff
@@ -0,0 +1,129 @@
+	}
+	if s.MikrotikIDField == "" {
+		return errors.New("Mikrotik ID field is not set to any of struct's fields")
+	}
```
Agree. Fully removed as the same error handling already part of parsing func (probably, these are left overs from initial implementation)

maksym-nazarenko commented on cmd/mikrotik-codegen/main.go:
```diff
@@ -0,0 +1,129 @@
+		return errors.New("Mikrotik ID field is not set to any of struct's fields")
+	}
+
+	if config.StructName == "" {
+		config.StructName = s.Name
+	}
```
No.
Initially, the intention was to fetch struct name from parsed `.go` file - the next `struct` definition after line with `//go:generate`.
However, you can explicitly pass the name as `mikrotik-codegen -struct MyStruct` to find specific struct.
At the end, you should read it as `if no struct name provided on command line, use one from parsed file`
Hope it makes sense.

ddelnano replied:
Yep, that does.

Could you add a comment here explaining that or add a comment struct to codegen.ParseFile function to indicate that its return value will find a viable struct from the file it parses? I don't think that is obvious from reading the code.

maksym-nazarenko replied:
updated comment for `ParseFile()` and added explanation here

maksym-nazarenko commented on go.mod:
```diff
@@ -5,10 +5,11 @@ go 1.18
 	github.com/hashicorp/terraform-plugin-docs v0.13.0
-	github.com/hashicorp/terraform-plugin-framework v1.0.1
-	github.com/hashicorp/terraform-plugin-go v0.14.2
+	github.com/hashicorp/terraform-plugin-framework v1.2.0
+	github.com/hashicorp/terraform-plugin-go v0.14.3
```
hmm, no - it wasn't intentional action 🤔

ddelnano commented on cmd/mikrotik-codegen/internal/codegen/templates.go:
```diff
@@ -0,0 +1,189 @@
+				Optional: {{.Optional}},
+				Computed: {{.Computed}},
+				{{if .Computed -}}
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
```
The usages for "string" in the struct and package name should be updated to use `.Type.Name`.

maksym-nazarenko replied:
done

ddelnano commented on cmd/mikrotik-codegen/internal/codegen/templates.go:
```diff
@@ -0,0 +1,189 @@
+
+// Metadata returns the resource type name.
+func (r *{{$resourceStructName}}) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_{{.ResourceName | lowercase}}"
```
I think this should use `ToSnakeCase` instead. I was expecting the `InterfaceWireguard` struct I tested with to result in `mikrotik_interface_wireguard`.

If that's not how you intended it to be used, can you explain how to form the existing resource names?

maksym-nazarenko replied:
nice catch!
I tested against `single word` resources and didn't even realize it would generate wrong name 😬

ddelnano commented:
I tried out this PR against the client and terraform test changes in #153. I didn't get completely through that testing, but I added comments for the things I uncovered so far.

From performing that testing, it made me wonder if we could test that the output of `go run ./cmd/mikrotik-codegen` results code that compiles. The `parser_test` seems to already leverage the go parser and AST, so maybe that wouldn't be too difficult to do?

It would be very useful to be able to add unit tests for a give struct definition to ensure things like `bool` and `int64` types use the correct `${TYPE}planmodifier.UseStateForUnknown()`. The verification I did so far was time consuming, so being able to have an end to end test for that would be useful. Curious to hear your thoughts on this.

maksym-nazarenko commented:
@ddelnano I addressed all the comments in the latest commits.
however, I would like to emphasize, that current implementation is not intended to be fully automatic.
I see at least these issues:
* it is not possible to use custom imports (e.g. if resource has no `Computed` field, `UseStateForUnknown()` is not used, so the import also should be skipped). I have something in my head, but I don't think I can implement it right away due to lack of time.
* For `int` field, the generator should wrap it in `int64()` and vice-versa for `modelTo...()` and `...ToModel()` funcs.
* Generator does not support custom types like `MikrotikDuration`

To fix types issues, I want to introduce some generic function that uses reflection, so it could be properly tested and just used in code generation.

To wrap-up: my intention is to ease the creation of new resources (by auto-generation boilerplate). Once we have enough use cases, we can proceed with enhancements.

ddelnano commented:
Yep, completely understand that this is meant to ease the creation of the boilerplate. The current implementation lgtm. Sorry for the long turn around time on the latest review :ship:
